tests/TestCase/Validation/ValidatorTest.php

```diff
@@ -706,6 +706,341 @@ public function testAllowEmptyAsArrayFailure()
         $validator->allowEmpty(['title' => 'derp', 'created' => false]);
     }
 
+    /**
+     * Tests the allowEmptyString method
+     *
+     * @return void
+     */
+    public function testAllowEmptyString()
+    {
+        $validator = new Validator();
+        $validator->allowEmptyString('title')
+            ->scalar('title');
+
+        $this->assertTrue($validator->isEmptyAllowed('title', true));
+        $this->assertTrue($validator->isEmptyAllowed('title', false));
+
+        $data = [
+            'title' => '',
+        ];
+        $this->assertEmpty($validator->errors($data));
+
+        $data = [
+            'title' => null,
+        ];
+        $this->assertEmpty($validator->errors($data));
+
+        $data = [
+            'title' => [],
+        ];
+        $this->assertNotEmpty($validator->errors($data));
+
+        $validator = new Validator();
+        $validator->allowEmptyString('title', 'update', 'message');
+        $this->assertFalse($validator->isEmptyAllowed('title', true));
+        $this->assertTrue($validator->isEmptyAllowed('title', false));
+
+        $data = [
+            'title' => null,
+        ];
+        $expected = [
+            'title' => ['_empty' => 'message'],
+        ];
+        $this->assertSame($expected, $validator->errors($data, true));
+        $this->assertEmpty($validator->errors($data, false));
+    }
+
+    /**
+     * Tests the allowEmptyArray method
+     *
+     * @return void
+     */
+    public function testAllowEmptyArray()
+    {
+        $validator = new Validator();
+        $validator->allowEmptyArray('items')
+            ->hasAtMost('items', 3);
+
+        $this->assertTrue($validator->field('items')->isEmptyAllowed());
+
+        $data = [
+            'items' => '',
+        ];
+        $result = $validator->errors($data);
+        $this->assertEmpty($result);
+
+        $data = [
+            'items' => null,
+        ];
+        $result = $validator->errors($data);
+        $this->assertEmpty($result);
+
+        $data = [
+            'items' => [],
+        ];
+        $result = $validator->errors($data);
+        $this->assertEmpty($result);
+
+        $data = [
+            'items' => [1, 2, 3, 4, 5],
+        ];
+        $result = $validator->errors($data);
+        $this->assertNotEmpty($result);
```

Comment:
Should this assert the message?

Reply:
What message should we report to users?

Reply:
Would it be useful to check that the `hasAtMost` rule failed?

Reply:
Ah sorry, I misunderstood. I will add the assertion to make sure hasAtMost rule failed.